src/ThermalShallowWaterExplicit.jl

```diff
@@ -0,0 +1,219 @@
+function compute_temperature_gradient!(dT,dω,V,RTMMchol,T)
+  b(v) = ∫(-1.0*DIV(v)*T)dω
+  Gridap.FESpaces.assemble_vector!(b, get_free_dof_values(dT), V)
+  ldiv!(RTMMchol, get_free_dof_values(dT))
+end
+
+function compute_buoyancy_flux!(eF,dΩ,V,RTMMchol,e,F)
+  b(v) = ∫((v⋅F)*e)dΩ
+  Gridap.FESpaces.assemble_vector!(b, get_free_dof_values(eF), V)
+  ldiv!(RTMMchol, get_free_dof_values(eF))
+end
+
+# assume that H1chol factorization has already been performed
+function compute_buoyancy!(e,dΩ,S,H1chol,E)
+  b(s) = ∫(s*E)dΩ
+  Gridap.FESpaces.assemble_vector!(b, get_free_dof_values(e), S)
+  ldiv!(H1chol, get_free_dof_values(e))
+end
+
+function compute_velocity_tswe!(u1,dΩ,dω,V,RTMMchol,u2,qAPVM,eAPVM,F,ϕ,dT,n,dt1,dt2)
+  b(v) = ∫(v⋅u2 - dt1*(qAPVM)*(v⋅⟂(F,n)) - dt1*(eAPVM)*v⋅dT)dΩ + ∫(dt2*DIV(v)*ϕ)dω
+  Gridap.FESpaces.assemble_vector!(b, get_free_dof_values(u1), V)
+  ldiv!(RTMMchol, get_free_dof_values(u1))
+end
+
+function thermal_shallow_water_explicit_time_step!(
+     h₂, u₂, E₂, hₚ, uₚ, Eₚ, ϕ, F, q₁, q₂, e₁, e₂, H1h, H1hchol, dT,  # in/out args
+     model, dΩ, dω, V, Q, R, S, f, h₁, u₁, E₁, hₘ, uₘ, Eₘ,            # in args
+     RTMMchol, L2MMchol, dt, τ, leap_frog)                            # more in args
+
+  # energetically balanced explicit second order thermal shallow water solver
+  #
+  # f          : coriolis force (field)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # E₁         : fluid buoyancy at current time level
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*h₁)
```

Comment:
To give the following a though (if u do not see it, forget it). 

Is it possible to recast `shallow_water_explicit_time_step!` such that  `thermal_shallow_water_explicit_time_step!` is able to reuse some code from the former? I see some common patterns among both solvers. It seems the thermal one is like some sort of extension to the regular one.

Reply:
This I don't know. Originally I wanted to pass in a function point to a method (explicit, IMEX, etc) to the time stepper, but this was inadequate as the different methods require different data, so I'm not sure how to generalise this...